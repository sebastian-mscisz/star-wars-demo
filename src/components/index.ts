export * from './character-details';
export * from './characters';
export * from './characters-card';
export * from './error-component';
export * from './items-with-pagination';
export * from './location-details';
export * from './locations';
export * from './locations-card';
export * from './marginal-components';
export * from './pagination-buttons';
export * from './spinner';
export * from './vehicle-details';
export * from './vehicles';
export * from './vehicles-card';
