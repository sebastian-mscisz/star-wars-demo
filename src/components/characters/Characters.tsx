import { withHandlers } from 'src/hoc';
import { CharactersUi } from './ui/CharactersUi';
import { useCharacters } from 'src/hooks';

const CharactersWithHandlers = withHandlers(CharactersUi);

export const Characters: React.FC = () => {
    const { data: characters, isLoading, loadingError } = useCharacters();

    return <CharactersWithHandlers isLoading={isLoading} characters={characters} hasError={!!loadingError} />;
};
