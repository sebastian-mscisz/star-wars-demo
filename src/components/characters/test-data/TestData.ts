import { Characters, Character, Species } from 'src/common';

export const TestCharacter: Character = {
    name: 'Character1',
    url: 'https://swapi.info/api/people/1',
    homeworld: 'https://swapi.info/api/planets/1',
    species: ['species/1'],
    vehicles: ['https://swapi.info/api/vehicles/1'],
};

export const TestCharacters: Characters = [
    TestCharacter,
    {
        name: 'Character2',
        url: 'https://swapi.info/api/people/2',
        vehicles: ['https://swapi.info/api/vehicles/1', 'https://swapi.info/api/vehicles/2'],
        homeworld: 'https://swapi.info/api/planets/2',
        species: ['species/2'],
    },
    {
        name: 'Character3',
        url: 'https://swapi.info/api/people/3',
        vehicles: [],
        homeworld: 'https://swapi.info/api/planets/3',
        species: ['species/3'],
    },
    {
        name: 'Character4',
        url: 'https://swapi.info/api/people/4',
        vehicles: ['https://swapi.info/api/vehicles/3'],
        homeworld: 'https://swapi.info/api/planets/4',
        species: ['species/4'],
    },
    {
        name: 'Character5',
        url: 'https://swapi.info/api/people/5',
        vehicles: ['https://swapi.info/api/vehicles/4', 'https://swapi.info/api/vehicles/5'],
        homeworld: 'https://swapi.info/api/planets/4',
        species: ['species/5'],
    },
];

export const TestSpecies: Species = [
    {
        name: 'Species2',
        url: 'species/2',
    },
    {
        name: 'Species2',
        url: 'species/2',
    },
    {
        name: 'Species3',
        url: 'species/3',
    },
    {
        name: 'Species4',
        url: 'species/4',
    },
    {
        name: 'Species5',
        url: 'species/5',
    },
];
