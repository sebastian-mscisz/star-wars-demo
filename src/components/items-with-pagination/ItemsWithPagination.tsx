import { Characters, Locations, Vehicles, notFoundPagePath } from 'src/common';
import { PaginationButtons } from 'src/components';
import { Navigate, Route, Routes } from 'react-router-dom';

type ItemsWithPaginationProps = {
    items: Locations | Characters | Vehicles;
    itemsPerPage: number;
    currentPageId: number;
    navPagePath: string;
    className: string;
    CardComponent: React.ElementType;
};

export const ItemsWithPagination: React.FC<ItemsWithPaginationProps> = ({
    items,
    itemsPerPage,
    currentPageId,
    navPagePath,
    className,
    CardComponent,
}) => {
    const itemsLine = items.map((item, i) => <CardComponent item={item} key={i} />);

    const numberOfPages = Math.floor(itemsLine.length / itemsPerPage);
    const itemsOnLastPage = itemsLine.length % itemsPerPage;

    const pageRoutes = Array(numberOfPages)
        .fill(null)
        .map((_, i) => i + 1)
        .map((el) => (
            <Route
                key={`${className}-page-${el + 1}`}
                path={`/${(el + 1).toString()}`}
                element={
                    <div className={`sw-${className}`}>{[...itemsLine].splice(el * itemsPerPage, itemsPerPage)}</div>
                }
            />
        ));

    return (
        <div className="sw-container__rows-wrapper">
            <PaginationButtons
                hideLeft={currentPageId === 1}
                hideRight={currentPageId === numberOfPages + (itemsOnLastPage ? 1 : 0)}
                leftPath={`${navPagePath}/${currentPageId - 1}`}
                rightPath={`${navPagePath}/${currentPageId + 1}`}
            />
            <Routes>
                {pageRoutes}
                <Route
                    index
                    element={<div className={`sw-${className}`}>{[...itemsLine].splice(0, itemsPerPage)}</div>}
                />
                <Route path="/1" element={<Navigate replace to={navPagePath} />} />
                <Route path="*" element={<Navigate replace to={notFoundPagePath()} />} />
            </Routes>
        </div>
    );
};
