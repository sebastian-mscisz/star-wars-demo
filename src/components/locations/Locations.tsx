import { withHandlers } from 'src/hoc';
import { LocationsUi } from './ui/LocationsUi';
import { useLocations } from 'src/hooks';

const LocationsWithHandlers = withHandlers(LocationsUi);

export const Locations: React.FC = () => {
    const { data: locations, isLoading, loadingError } = useLocations();

    return <LocationsWithHandlers isLoading={isLoading} locations={locations} hasError={!!loadingError} />;
};
