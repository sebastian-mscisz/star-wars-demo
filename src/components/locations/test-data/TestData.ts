import { Locations, Location } from 'src/common';

export const TestLocation: Location = {
    name: 'Location1',
    url: 'https://swapi.info/api/planets/1',
    residents: ['https://swapi.info/api/people/1'],
    population: '1000',
};

export const TestLocations: Locations = [
    TestLocation,
    {
        name: 'Location2',
        url: 'https://swapi.info/api/planets/2',
        residents: ['https://swapi.info/api/people/2'],
        population: '5343',
    },
    {
        name: 'Location3',
        url: 'https://swapi.info/api/planets/3',
        residents: ['https://swapi.info/api/people/3'],
        population: '13333',
    },
    {
        name: 'Location4',
        url: 'https://swapi.info/api/planets/4',
        residents: ['https://swapi.info/api/people/4', 'https://swapi.info/api/people/5'],
        population: '9999',
    },
    {
        name: 'Location5',
        url: 'https://swapi.info/api/planets/5',
        residents: [],
        population: '999',
    },
];
