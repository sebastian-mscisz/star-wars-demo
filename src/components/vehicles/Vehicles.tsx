import { withHandlers } from 'src/hoc';
import { VehiclesUi } from './ui/VehiclesUi';
import { useVehicles } from 'src/hooks';

const VehiclesWithHandlers = withHandlers(VehiclesUi);

export const Vehicles: React.FC = () => {
    const { data: vehicles, isLoading, loadingError } = useVehicles();

    return <VehiclesWithHandlers isLoading={isLoading} vehicles={vehicles} hasError={!!loadingError} />;
};
