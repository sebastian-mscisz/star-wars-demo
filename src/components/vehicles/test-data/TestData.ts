import { Vehicles, Vehicle } from 'src/common';

export const TestVehicle: Vehicle = {
    name: 'Vehicle1',
    url: 'https://swapi.info/api/vehicles/1',
    vehicle_class: 'Air speeder',
    pilots: ['character/1', 'character/2'],
};

export const TestVehicles: Vehicles = [
    TestVehicle,
    {
        name: 'Vehicle2',
        url: 'https://swapi.info/api/vehicles/2',
        pilots: ['character/2'],
        vehicle_class: 'Land speeder',
    },
    {
        name: 'Vehicle3',
        url: 'https://swapi.info/api/vehicles/3',
        pilots: ['character/4'],
        vehicle_class: 'Water speeder',
    },
    {
        name: 'Vehicle4',
        url: 'https://swapi.info/api/vehicles/4',
        pilots: ['character/5'],
        vehicle_class: 'Fire speeder',
    },
    {
        name: 'Vehicle5',
        url: 'https://swapi.info/api/vehicles/5',
        pilots: ['character/5'],
        vehicle_class: 'Air speeder',
    },
];
