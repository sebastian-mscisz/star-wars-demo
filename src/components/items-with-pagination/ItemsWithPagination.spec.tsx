import { BrowserRouter } from 'react-router-dom';
import { ItemsWithPagination } from './ItemsWithPagination';
import { render } from '@testing-library/react';
import { TestVehicles } from '../vehicles/test-data';
import { TestCharacters } from '../characters/test-data';
import { TestLocations } from '../locations/test-data';
import { charactersPagePath, locationsPagePath, vehiclesPagePath } from 'src/common';
import { VehiclesCard, LocationsCard, CharactersCard } from 'src/components';

describe('ItemsWithPagination', () => {
    it('should render with vehicles', () => {
        const { asFragment } = render(
            <ItemsWithPagination
                itemsPerPage={4}
                currentPageId={1}
                className="vehicles"
                CardComponent={VehiclesCard}
                items={TestVehicles}
                navPagePath={vehiclesPagePath()}
            />,
            { wrapper: BrowserRouter },
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('should render with characters', () => {
        const { asFragment } = render(
            <ItemsWithPagination
                itemsPerPage={4}
                currentPageId={1}
                className="characters"
                CardComponent={CharactersCard}
                items={TestCharacters}
                navPagePath={charactersPagePath()}
            />,
            { wrapper: BrowserRouter },
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('should render with locations', () => {
        const { asFragment } = render(
            <ItemsWithPagination
                itemsPerPage={4}
                currentPageId={1}
                className="locations"
                CardComponent={LocationsCard}
                items={TestLocations}
                navPagePath={locationsPagePath()}
            />,
            { wrapper: BrowserRouter },
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
