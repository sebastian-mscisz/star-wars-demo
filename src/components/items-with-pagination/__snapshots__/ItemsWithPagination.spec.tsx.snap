// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemsWithPagination should render with characters 1`] = `
<DocumentFragment>
  <div
    class="sw-pagination-section"
    data-testid="pagination-section"
  >
    <div
      class="sw-pagination-button sw-pagination-button--hidden"
      data-testid="left-pagination-button"
    >
      ←
    </div>
    <div
      class="sw-pagination-button "
      data-testid="right-pagination-button"
    >
      →
    </div>
  </div>
  <div
    class="sw-characters"
  >
    <div
      class="sw-characters-card"
    >
      <div
        class="sw-characters-card__tile"
      >
        <a
          class="sw-characters-card__tile__link"
          href="/character/1"
        >
          <div
            class="sw-characters-card__tile__link__image-wrapper"
          >
            <img
              alt="Star wars character"
              loading="lazy"
              src="/characters/1.jpg"
            />
          </div>
          <h5
            class="sw-characters-card__tile__link__text"
          >
            Character1
          </h5>
        </a>
      </div>
    </div>
    <div
      class="sw-characters-card"
    >
      <div
        class="sw-characters-card__tile"
      >
        <a
          class="sw-characters-card__tile__link"
          href="/character/2"
        >
          <div
            class="sw-characters-card__tile__link__image-wrapper"
          >
            <img
              alt="Star wars character"
              loading="lazy"
              src="/characters/2.jpg"
            />
          </div>
          <h5
            class="sw-characters-card__tile__link__text"
          >
            Character2
          </h5>
        </a>
      </div>
    </div>
    <div
      class="sw-characters-card"
    >
      <div
        class="sw-characters-card__tile"
      >
        <a
          class="sw-characters-card__tile__link"
          href="/character/3"
        >
          <div
            class="sw-characters-card__tile__link__image-wrapper"
          >
            <img
              alt="Star wars character"
              loading="lazy"
              src="/characters/3.jpg"
            />
          </div>
          <h5
            class="sw-characters-card__tile__link__text"
          >
            Character3
          </h5>
        </a>
      </div>
    </div>
    <div
      class="sw-characters-card"
    >
      <div
        class="sw-characters-card__tile"
      >
        <a
          class="sw-characters-card__tile__link"
          href="/character/4"
        >
          <div
            class="sw-characters-card__tile__link__image-wrapper"
          >
            <img
              alt="Star wars character"
              loading="lazy"
              src="/characters/4.jpg"
            />
          </div>
          <h5
            class="sw-characters-card__tile__link__text"
          >
            Character4
          </h5>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ItemsWithPagination should render with locations 1`] = `
<DocumentFragment>
  <div
    class="sw-pagination-section"
    data-testid="pagination-section"
  >
    <div
      class="sw-pagination-button sw-pagination-button--hidden"
      data-testid="left-pagination-button"
    >
      ←
    </div>
    <div
      class="sw-pagination-button "
      data-testid="right-pagination-button"
    >
      →
    </div>
  </div>
  <div
    class="sw-locations"
  >
    <div
      class="sw-locations-card"
    >
      <div
        class="sw-locations-card__tile"
      >
        <a
          class="sw-locations-card__tile__link"
          href="/location/1"
        >
          <h5
            class="sw-locations-card__tile__link__text"
          >
            Location1
          </h5>
          <div
            class="sw-locations-card__tile__link__image-wrapper"
          >
            <img
              alt="Star wars planet"
              src="/locations/1.jpg"
            />
          </div>
        </a>
      </div>
    </div>
    <div
      class="sw-locations-card"
    >
      <div
        class="sw-locations-card__tile"
      >
        <a
          class="sw-locations-card__tile__link"
          href="/location/2"
        >
          <h5
            class="sw-locations-card__tile__link__text"
          >
            Location2
          </h5>
          <div
            class="sw-locations-card__tile__link__image-wrapper"
          >
            <img
              alt="Star wars planet"
              src="/locations/2.jpg"
            />
          </div>
        </a>
      </div>
    </div>
    <div
      class="sw-locations-card"
    >
      <div
        class="sw-locations-card__tile"
      >
        <a
          class="sw-locations-card__tile__link"
          href="/location/3"
        >
          <h5
            class="sw-locations-card__tile__link__text"
          >
            Location3
          </h5>
          <div
            class="sw-locations-card__tile__link__image-wrapper"
          >
            <img
              alt="Star wars planet"
              src="/locations/3.jpg"
            />
          </div>
        </a>
      </div>
    </div>
    <div
      class="sw-locations-card"
    >
      <div
        class="sw-locations-card__tile"
      >
        <a
          class="sw-locations-card__tile__link"
          href="/location/4"
        >
          <h5
            class="sw-locations-card__tile__link__text"
          >
            Location4
          </h5>
          <div
            class="sw-locations-card__tile__link__image-wrapper"
          >
            <img
              alt="Star wars planet"
              src="/locations/4.jpg"
            />
          </div>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ItemsWithPagination should render with vehicles 1`] = `
<DocumentFragment>
  <div
    class="sw-pagination-section"
    data-testid="pagination-section"
  >
    <div
      class="sw-pagination-button sw-pagination-button--hidden"
      data-testid="left-pagination-button"
    >
      ←
    </div>
    <div
      class="sw-pagination-button "
      data-testid="right-pagination-button"
    >
      →
    </div>
  </div>
  <div
    class="sw-vehicles"
  >
    <div
      class="sw-vehicles-card"
    >
      <div
        class="sw-vehicles-card__tile"
      >
        <a
          class="sw-vehicles-card__tile__link"
          href="/vehicle/1"
        >
          <div
            class="sw-vehicles-card__tile__link__image-wrapper"
          >
            <img
              alt="Star wars vehicle"
              src="/vehicles/1.jpg"
            />
          </div>
          <h5
            class="sw-vehicles-card__tile__link__text"
          >
            Vehicle1
          </h5>
        </a>
      </div>
    </div>
    <div
      class="sw-vehicles-card"
    >
      <div
        class="sw-vehicles-card__tile"
      >
        <a
          class="sw-vehicles-card__tile__link"
          href="/vehicle/2"
        >
          <div
            class="sw-vehicles-card__tile__link__image-wrapper"
          >
            <img
              alt="Star wars vehicle"
              src="/vehicles/2.jpg"
            />
          </div>
          <h5
            class="sw-vehicles-card__tile__link__text"
          >
            Vehicle2
          </h5>
        </a>
      </div>
    </div>
    <div
      class="sw-vehicles-card"
    >
      <div
        class="sw-vehicles-card__tile"
      >
        <a
          class="sw-vehicles-card__tile__link"
          href="/vehicle/3"
        >
          <div
            class="sw-vehicles-card__tile__link__image-wrapper"
          >
            <img
              alt="Star wars vehicle"
              src="/vehicles/3.jpg"
            />
          </div>
          <h5
            class="sw-vehicles-card__tile__link__text"
          >
            Vehicle3
          </h5>
        </a>
      </div>
    </div>
    <div
      class="sw-vehicles-card"
    >
      <div
        class="sw-vehicles-card__tile"
      >
        <a
          class="sw-vehicles-card__tile__link"
          href="/vehicle/4"
        >
          <div
            class="sw-vehicles-card__tile__link__image-wrapper"
          >
            <img
              alt="Star wars vehicle"
              src="/vehicles/4.jpg"
            />
          </div>
          <h5
            class="sw-vehicles-card__tile__link__text"
          >
            Vehicle4
          </h5>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
